[RenderStyleGen] Generate getter/setter function declarations

weinig opened the pull request:
#### b7a21d91ed7997f9ddac8ae0ce9e5d72cdb4c07f
<pre>
[RenderStyleGen] Generate getter/setter function declarations
<a href="https://bugs.webkit.org/show_bug.cgi?id=303356">https://bugs.webkit.org/show_bug.cgi?id=303356</a>

Reviewed by Darin Adler.

To generate the declarations for RenderStyle&apos;s property getter/setters,
we have to split RenderStyle, ending up with three classes:

    - RenderStyleBase: hand written, base class, contains member variables and non-property getter/setters
    - RenderStyleProperties: generated, derives from RenderStyleBase, contains the property getter/setters
    - RenderStyle: hand written, derives from RenderStyleProperties, contains logic that depends on property getter/setters

I am not sure I have the exact right balance about what is in RenderStyleBase
vs RenderStyle, but that can be tweaked going forward.

* Source/WebCore/Headers.cmake:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleInlinesGenerated.h: Removed.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties.h: Added.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStylePropertiesGetters.h: Added.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStylePropertiesSetters.h: Copied from Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleSettersGenerated.h.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStylePropertiesSettersInlines.h: Renamed from Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleSettersGenerated.h.
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/OutlineValue.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleBase.cpp: Added.
* Source/WebCore/rendering/style/RenderStyleBase.h: Added.
* Source/WebCore/rendering/style/RenderStyleBaseInlines.h: Added.
* Source/WebCore/rendering/style/RenderStyleBaseSetters.h: Added.
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.cpp: Added.
* Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.h: Added.
* Source/WebCore/rendering/style/RenderStylePropertiesSettersCustom.h: Added.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.h:
* Source/WebCore/style/values/css2/StyleZIndex.h:
* Source/WebCore/style/values/masking/StyleMaskBorder.h:
* Source/WebCore/style/values/svg/StyleSVGPaint.cpp:

Canonical link: <a href="https://commits.webkit.org/304026@main">https://commits.webkit.org/304026@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9010c4e9ebab5872d299d820a13922f5408f11c1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134281 "15 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6788 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45502 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141858 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7370 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6651 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102677 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137228 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5141 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120366 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83470 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/133630 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2627 "Passed tests") | [⏳ 🛠 wpe-cairo-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-Cairo-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38503 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144508 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6459 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39082 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111063 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6546 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5436 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111314 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28241 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4852 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116635 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60247 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6511 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34847 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6346 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70066 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6585 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6459 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -109,15 +108,122 @@ struct SameSizeAsRenderStyle : CanMakeCheckedPtr<SameSizeAsRenderStyle> {
+    m_inheritedFlags.textAlign = static_cast<unsigned>(RenderStyle::initialTextAlign());
+    m_inheritedFlags.textTransform = RenderStyle::initialTextTransform().toRaw();
+    m_inheritedFlags.textDecorationLineInEffect = RenderStyle::initialTextDecorationLine().toRaw();
+    m_inheritedFlags.cursorType = static_cast<unsigned>(RenderStyle::initialCursor().predefined);
+#if ENABLE(CURSOR_VISIBILITY)
+    m_inheritedFlags.cursorVisibility = static_cast<unsigned>(RenderStyle::initialCursorVisibility());
+#endif
+    m_inheritedFlags.whiteSpaceCollapse = static_cast<unsigned>(RenderStyle::initialWhiteSpaceCollapse());
+    m_inheritedFlags.textWrapMode = static_cast<unsigned>(RenderStyle::initialTextWrapMode());
+    m_inheritedFlags.textWrapStyle = static_cast<unsigned>(RenderStyle::initialTextWrapStyle());
+    m_inheritedFlags.borderCollapse = static_cast<unsigned>(RenderStyle::initialBorderCollapse());
+    m_inheritedFlags.rtlOrdering = static_cast<unsigned>(RenderStyle::initialRTLOrdering());
+    m_inheritedFlags.boxDirection = static_cast<unsigned>(RenderStyle::initialBoxDirection());
+    m_inheritedFlags.printColorAdjust = static_cast<unsigned>(RenderStyle::initialPrintColorAdjust());
+    m_inheritedFlags.pointerEvents = static_cast<unsigned>(RenderStyle::initialPointerEvents());
+    m_inheritedFlags.insideLink = static_cast<unsigned>(InsideLink::NotInside);
+#if ENABLE(TEXT_AUTOSIZING)
+    m_inheritedFlags.autosizeStatus = 0;
+#endif
+
+    m_nonInheritedFlags.effectiveDisplay = static_cast<unsigned>(RenderStyle::initialDisplay());
+    m_nonInheritedFlags.originalDisplay = static_cast<unsigned>(RenderStyle::initialDisplay());
+    m_nonInheritedFlags.overflowX = static_cast<unsigned>(RenderStyle::initialOverflowX());
+    m_nonInheritedFlags.overflowY = static_cast<unsigned>(RenderStyle::initialOverflowY());
+    m_nonInheritedFlags.clear = static_cast<unsigned>(RenderStyle::initialClear());
+    m_nonInheritedFlags.position = static_cast<unsigned>(RenderStyle::initialPosition());
+    m_nonInheritedFlags.unicodeBidi = static_cast<unsigned>(RenderStyle::initialUnicodeBidi());
+    m_nonInheritedFlags.floating = static_cast<unsigned>(RenderStyle::initialFloating());
+    m_nonInheritedFlags.textDecorationLine = RenderStyle::initialTextDecorationLine().toRaw();
+    m_nonInheritedFlags.usesViewportUnits = false;
+    m_nonInheritedFlags.usesContainerUnits = false;
+    m_nonInheritedFlags.useTreeCountingFunctions = false;
+    m_nonInheritedFlags.hasExplicitlyInheritedProperties = false;
+    m_nonInheritedFlags.disallowsFastPathInheritance = false;
+    m_nonInheritedFlags.emptyState = false;
+    m_nonInheritedFlags.firstChildState = false;
+    m_nonInheritedFlags.lastChildState = false;
+    m_nonInheritedFlags.isLink = false;
+    m_nonInheritedFlags.pseudoElementType = 0;
+    m_nonInheritedFlags.pseudoBits = 0;
```
Any way to avoid or minimize this repetitive code?

weinig replied:
Seems like it. Now that we have the storage paths, we should be able to generate some initialization functions for all the property values in that storage, but I will leave that for a future change.

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -109,15 +108,122 @@ struct SameSizeAsRenderStyle : CanMakeCheckedPtr<SameSizeAsRenderStyle> {
+
+inline void RenderStyleBase::NonInheritedFlags::copyNonInheritedFrom(const NonInheritedFlags& other)
+{
+    // Only some flags are copied because NonInheritedFlags contains things that are not actually style data.
```
Is there a way to make this less risky, less chance of missing one here or including one that shouldn’t? Granted, this is not new, and was just moved, so maybe not so relevant to this pull request.

weinig replied:
Something to think about for sure though.

weinig committed:
[RenderStyleGen] Generate getter/setter function declarations
https://bugs.webkit.org/show_bug.cgi?id=303356

Reviewed by Darin Adler.

To generate the declarations for RenderStyle's property getter/setters,
we have to split RenderStyle, ending up with three classes:

    - RenderStyleBase: hand written, base class, contains member variables and non-property getter/setters
    - RenderStyleProperties: generated, derives from RenderStyleBase, contains the property getter/setters
    - RenderStyle: hand written, derives from RenderStyleProperties, contains logic that depends on property getter/setters

I am not sure I have the exact right balance about what is in RenderStyleBase
vs RenderStyle, but that can be tweaked going forward.

* Source/WebCore/Headers.cmake:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleInlinesGenerated.h: Removed.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties.h: Added.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStylePropertiesGetters.h: Added.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStylePropertiesSetters.h: Copied from Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleSettersGenerated.h.
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStylePropertiesSettersInlines.h: Renamed from Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleSettersGenerated.h.
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/OutlineValue.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleBase.cpp: Added.
* Source/WebCore/rendering/style/RenderStyleBase.h: Added.
* Source/WebCore/rendering/style/RenderStyleBaseInlines.h: Added.
* Source/WebCore/rendering/style/RenderStyleBaseSetters.h: Added.
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.cpp: Added.
* Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.h: Added.
* Source/WebCore/rendering/style/RenderStylePropertiesSettersCustom.h: Added.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.h:
* Source/WebCore/style/values/css2/StyleZIndex.h:
* Source/WebCore/style/values/masking/StyleMaskBorder.h:
* Source/WebCore/style/values/svg/StyleSVGPaint.cpp:

Canonical link: https://commits.webkit.org/304026@main
